File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.java
class Solution {
    public int longestSubsequence(int[] arr, int diff) {
        HashMap<Integer, Integer> dp = new HashMap<>();
        int ans = 1;
    
        for(int el: arr) {
            if(dp.containsKey(el - diff)) {
                int prev = dp.get(el - diff);
                ans = Math.max(ans, 1+prev);
                dp.put(el, prev+1);
            } else {
                dp.put(el, 1);
            }
        }
        
        return ans;
    }
}

File: 1218-longest-arithmetic-subsequence-of-given-difference/1218-longest-arithmetic-subsequence-of-given-difference.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        unordered_map<int, int> dp;
        int ans = 1;
        
        for(int &el: arr) {
            if(dp.count(el - difference)) {
                int prev = dp[el - difference];
                ans = max(ans, prev+1);
                dp[el] = prev+1;
            } else {
                dp[el] = 1;
            }
        }
        
        return ans;
    }
};